[DRAFT] Initial version of the boundary event documentation (#1)
DmytroKost/docs: content/en/docs/refguide/modeling/application-logic/workflows/boundary-events.md

## Introduction

* [BPMN 2.0](#BPMN-2.0)
* [Types](#types)
* [Supporting Activities](#supporting-activities)
* [Getting started](#getting-started)
* [Adding Timer Boundary Events](#adding-timer-boundary-events)
* [Rearranging](#rearranging)
* [Execution](#execution)
* [Current limitations](#current-limitations)
* [Read More](#read-more)

## BMPN 2.0

In BPMN 2.0, boundary events are a type of event that is attached to the boundary of an activity (such as a task or subprocess) to handle exceptional situations or are triggered by certain behaviours. The icon inside the event indicates the type of event (e.g. clock indicates it’s a timer event). Boundary Events are always displayed by 2 circles (either solid or dashed) and are linked by a dotted line to the parent activity. 

There are two main types of boundary events:

- Non-Interrupting Boundary Events: These events do not interrupt the ongoing activity. When triggered, they allow the activity to continue while simultaneously starting a new path from the boundary event. As per BPMN2.0 specification, non-interrupting boundary events are visualised as two dashed circles with an icon in the center.

> - Interrupting Boundary Events: When these events are triggered, they interrupt the normal path of the activity they are attached to. The activity stops, and the process flow is redirected to the boundary event's outgoing sequence path. As per BPMN2.0 specification, interrupting boundary events are visualised as two solid circles.

Reviewer: pls add that interrupting not part of beta.

Author: Added extra line which explains this.

| {{< figure src="/attachments/refguide/modeling/application-logic/workflows/workflow-elements/call-microflow.png" alt="Call Microflow" class="no-border" >}} | [Call microflow](/refguide/call-microflow/)
| {{< figure src="/attachments/refguide/modeling/application-logic/workflows/workflow-elements/call-workflow.png" alt="Call Workflow" class="no-border" >}} | [Call workflow](/refguide/call-workflow/)

## Getting started

To get started with the new boundary events feature, these can be enabled by opening Studio Pro's settings and navigating to new features (this setting is only applicable during the beta).

- Windows: Preferences -> New Features -> Enable non-interrupting timer boundary events (beta)
- Mac (Beta): Preferences -> New Features -> Enable non-interrupting timer boundary events (beta)

{{< figure src="/attachments/refguide/modeling/application-logic/workflows/workflow-how-to-boundary-events/boundary-events-experimental-option.png" alt="Experimental option" width="450" class="no-border">}}

> ## Adding Timer Boundary Events

Reviewer: Since we will have multiple boundary event types in the future does it make sense to make this section more generic for all boundary events?
The first 2 steps are quit generic. Only the last step is specific for timer events, We would leave that for now or also refer to the specific event configuration page (Timer in this case)
@Yiyun333 WDYT?

Reviewer 2: @PaulJongenMx This section should be more generic I think. I will say "Adding a Boundary Event" in this case. The instructions should be generic as well and when needed, we can use Timer as an example. 

Please see my suggestion below: 

> ## Adding a Boundary Event
> 
> To add a boundary event to the above-listed activities, choose one of the following ways:
> 
> * Select an event from the **Events** section in the workflow **Toolbox** (currently only Timer is available) and drag it onto one of the activities listed above. Double-click the event to open its properties dialog box.
> 
> {{< figure src="/attachments/refguide/modeling/application-logic/workflows/workflow-how-to-boundary-events/drag-drop.png" alt="Adding boundary events through drag and drop" width="200" class="no-border">}}
> 
> * Double-click one of the above-listed activities to open its properties dialog box, go to the **Events** tab, and in the > **Boundary events** section, click **New** to add a new boundary event (currently only Timer is available).
> 
> {{< figure src="/attachments/refguide/modeling/application-logic/workflows/workflow-how-to-boundary-events/edit-dialog.png" alt="Adding boundary events through edit dialog" width="450" class="no-border">}}
> 
> For more information on how to configure the properties of a timer boundary event, see the [Properties](/refguide/timer/#properties) section in *Timer*.

I removed all the numbers in the instructions. I don't think they are needed because they can be confusing too. I can remove them from the screenshot as well.

Author: I used yiyun's suggestion

2. Open the activity (e.g. User Task) properties (1), go to the Events tab (2) and create a new boundary event (3). Select the Timer event (4).

{{< figure src="/attachments/refguide/modeling/application-logic/workflows/workflow-how-to-boundary-events/edit-dialog.png" alt="Adding boundary events through edit dialog" width="450" class="no-border">}}

The next step is to configure the boundary timer properties like Caption and timer Duration. This could be a fixed period (after the parent task is initiated) or a date+time can be defined in an expression. For more information see [Timer Event](/refguide/timer-event/).

## Rearranging

You can change the ordering of the boundary event in 2 ways:

1. Via the context menu (right mouse click) of the timer event and select **Move boundary event left** or **Move boundary event right**, or by using the keyboard shortcuts [Keyboard Shortcuts](/refguide/keyboard-shortcuts/#editors-only).
 
{{< figure src="/attachments/refguide/modeling/application-logic/workflows/boundary-events/arrange-in-editor.png" width="350px" alt="Timer Event arrange in editor" class="no-border" >}}

> 2. Open the edit dialog of the parent activity and change the order of the boundary event paths by altering the list. Note that this does not change the order execution of the paths, as this is dependent on the expiration of the timer

Reviewer 3: Maybe extract `Note that this does not change the order execution of the paths, as this is dependent on the expiration of the timer` as a separate line instead of being part of the 2nd bullet point.

Reviewer 2: I'll put this in an info alert under the 2nd bullet point.

Author: I applied yiyun's suggestion

{{< figure src="/attachments/refguide/modeling/application-logic/workflows/boundary-events/arrange-in-dialog.png" width="450px" alt="Arrange non-interrupting boundary event in dialog" class="no-border" >}}

> ## Execution

Reviewer: We should have this as well in the Workflow Engine page. This info is useful in both locations.

Author: I added this information to the engine page as well

Boundary events are initiated when their parent activity is initiated. For example, for a timer with a fixed duration, it will start its count down when the parent activity is initiated. In case the parent activity is completed before any of the boundary events are triggered, none of the activities in the boundary event path will be executed and all timers will be cancelled.

With non-interrupting boundary events the parent activity remains active/in progress when an event is triggered (i.e., the parent activity is not interrupted). For example, when a timer event on a user task is triggered after 2 days, this task will remain in progress and the path defined below the timer event is executed. When the boundary event path reaches the **End of Boundary Path** activity, the workflow will await the completion of the parent activity. 

## Current Limitations

> 

Reviewer: @Yiyun333 Is it OK to have these mentioned here? In particular the first limitation is something most developers will run into and wondering what they have mist to make it possible.

Reviewer 2: @PaulJongenMx Yes, I think it is okay for now to list those limitations here.

Author: Kept the limitations in place

The current release (10.15) of boundary events has the following functional limitations (which are actively being developed):

> - No access to data of the parent activity in the boundary path. For user tasks we will soon add a variable **$ParentTask** which is available in the boundary event path of a user task (e.g. do get the assigned user that should receive a notification when a task is overdue). Same for $CalledWorkflowInstance in case the parent activity is a Call workflow activity.

Reviewer 3: `(e.g. do get the assigned user that should receive a notification when a task is overdue)` is a weird sentence

Author: Changed the sentence up a bit, to make it a bit more understandable and in line.

> - Non-interrupting timer boundary events have no recurrence (they are only executed once and will not repeat). This attribute will be added in a future update.

Reviewer 3: Do we want to specify that it is coming in a future update? It think it's better to just say we don't have it right now.

```suggestion
- Non-interrupting timer boundary events currently have no recurrence (they are only executed once and will not repeat).
```

Author: Changed it to the suggestion.

---
title: "Boundary Events"
url: /refguide/boundary-events/
weight: 55
---

## Introduction

> Starting with Mendix version 10.15, the workflow editor has beta support for Non-Interrupting Timer boundary events.

Reviewer 3: What's the rule / strategy for capitalization here regarding `Non-Interrupting Timer boundary events.`

Examples:

- Send a reminder when a task is not picked up after 2 days.
- Escalate to a manager when a task is not completed after 10 days.

## Overview

> In BPMN 2.0, boundary events are a type of event that is attached to the boundary of an activity (such as a task or subprocess) to handle exceptional situations or to trigger certain behaviours. The icon inside the event indicates the type of event (e.g. clock indicates it’s a timer event). Boundary events are always displayed by 2 circles (either solid or dashed) and are linked by a dotted line to the parent activity.

Reviewer 3: Maybe also explain what the difference between an solid and dashed line is because we'll have both when we introduce interrupting events

{{< figure src="/attachments/refguide/modeling/application-logic/workflows/boundary-events/boundary-event.png" max-width=100% >}}

A boundary event can be added to the following activities:

- User Task
- Multi-User Task
- Wait for Notification
- Call Workflow
- Call Microflow

A maximum of 5 boundary events can be added to an activity.

> Boundary events are intermediate events of an activity, which can be triggered while the related activity is in progress. When the boundary event is triggered its effect on the related activity depends on the boundary event type: Non-Interrupting or Interrupting boundary event.

Reviewer 3: Please explain (or refer to) what intermediate means.

Reviewer 4: That's a term from BPMN, not sure we should mention it here. What do you think @PaulJongenMx ?

Reviewer 3: If we don't add anything, I think it will be a bit confusing. Especially for new users or people who are unfamiliar with BPMN. Maybe we can refer to an external source?

### Non-Interrupting Boundary Event {#non-interrupting}

Non-Interrupting boundary events do not interrupt the ongoing activity. When triggered, they allow the activity to continue while simultaneously starting a new flow from the boundary event. As per BPMN2.0 specification, non-interrupting boundary events are visualised as two dashed circles with an icon in the center.

### Interrupting Boundary Event {#interrupting}

When interrupting boundary events are triggered, they interrupt the normal flow of the activity they are attached to. The activity stops, and the process flow is redirected to the boundary event's outgoing sequence flow. As per BPMN2.0 specification, interrupting boundary events are visualised as two solid circles. Interrupting boundary Events are not currently supported by the workflow engine. We have plans to release them in the future.

## Boundary Event Types

There are several types of boundray events depending on the condition that defines how and when the boundary event is triggered. Currently, the workflow engine supports Timer boundary events.

### Timer Boundary Event {#timer}

> A timer boundary event is executed with the specified delay. When the activity execution starts, it shedules a corresponding timer. When the corresponding timer expires the boundary event flow begins its execution. For more information on how the **Workflow Engine** handles boundary events see [Timer Boundary Events](/refguide/workflow-engine/#timer-boundary-events) section.

Reviewer 3: ```suggestion
A timer boundary event is executed with the specified delay. When the activity execution starts, it shedules a corresponding timer. When this timer expires, the boundary event flow begins its execution. For more information on how the **Workflow Engine** handles boundary events see [Timer Boundary Events](/refguide/workflow-engine/#timer-boundary-events) section.
```

There are two ways to add a timer boundary event to an activity:

1. Drag **Timer** from the toolbox onto the activity that supports boundary events. {{< figure src="/attachments/refguide/modeling/application-logic/workflows/boundary-events/timer-toolbox.png" max-width=100% >}}

> 1. Open the activity properties, go to the **Events** tab, create a new boudary event by clicking **New** and select the **Timer** boundary event type. {{< figure src="/attachments/refguide/modeling/application-logic/workflows/boundary-events/timer-events-tab.png" max-width=100% >}}

Reviewer 3: ```suggestion
1. Open its properties, go to the **Events** tab, create a new boundary event by clicking **New** and select the **Timer** boundary event type. {{< figure src="/attachments/refguide/modeling/application-logic/workflows/boundary-events/timer-events-tab.png" max-width=100% >}}
```

> A Timer Boundary Event must have either a timer duration or a timer expression. They can be configured in the [Timer](/refguide/timer/#timer) section. The configuration properties can be accessed by either double-cliking on the **Timer** icon or by going to the activity properties **Events** tab and double clicking on the **Timer** event from the list.

Reviewer 3: ```suggestion
A Timer Boundary Event must have either a timer duration or a timer expression. This can be configured in the [Timer](/refguide/timer/#timer) section. The configuration properties can be accessed by either double-clicking on the **Timer** icon on the canvas or by navigating to the  **Events** tab in the activity's properties and double clicking the **Timer** event from the list.
```
